data_steward/cdr_cleaner/clean_cdr_engine.py

```diff
@@ -140,6 +149,226 @@ def clean_dataset(project=None, statements=None, data_stage=stage.UNSPECIFIED):
         )
 
 
+def get_destination_ref(project_id, query_dict):
+    """
+    Fetch BQ table reference to destination table
+
+    :param project_id: Identifies the project
+    :param query_dict: dictionary for the query
+    :return: BQ TableReference object to destination
+    """
+    destination_dataset_id = query_dict.get(cdr_consts.DESTINATION_DATASET)
+    destination_dataset_ref = gbq.DatasetReference(project_id,
+                                                   destination_dataset_id)
+    destination_table_id = query_dict.get(cdr_consts.DESTINATION_TABLE)
+    destination_table_ref = gbq.table.TableReference(destination_dataset_ref,
+                                                     destination_table_id)
+    return destination_table_ref
+
+
+def generate_job_config(project_id, query_dict):
+    """
+    Generates BigQuery job_configuration object
+
+    :param project_id: Identifies the project
+    :param query_dict: dictionary for the query
+    :return: BQ job_configuration object
+    """
+    job_config = gbq.job.QueryJobConfig()
+    if query_dict.get(cdr_consts.DESTINATION_TABLE) is None:
+        return job_config
+
+    destination_table = get_destination_ref(project_id, query_dict)
+
+    job_config.destination = destination_table
+    job_config.use_legacy_sql = query_dict.get(cdr_consts.LEGACY_SQL, False)
+    # allow_large_results can only be used if use_legacy_sql=True
+    job_config.allow_large_results = job_config.use_legacy_sql
+    job_config.write_disposition = query_dict.get(cdr_consts.DISPOSITION,
+                                                  bq_consts.WRITE_EMPTY)
+    return job_config
+
+
+def run_queries(client, query_list, rule_info):
+    """
+    Runs queries from the list of query_dicts
+
+    :param client: BigQuery client
+    :param query_list: list of query_dicts generated by a cleaning rule
+    :param rule_info: contains information about the query function
+    :return: integers indicating the number of queries that succeeded and failed
+    """
+    query_count = len(query_list)
+    successes, failures = [], []
+    for index, query_dict in enumerate(query_list):
+        try:
+            LOGGER.info(
+                ce_consts.QUERY_RUN_MESSAGE_TEMPLATE.render(
+                    index, query_count, **rule_info, **query_dict))
+            job_config = generate_job_config(client.project, query_dict)
+
+            module_name = rule_info[cdr_consts.MODULE_NAME]
+            query_job = client.query(query=query_dict.get(cdr_consts.QUERY),
+                                     job_config=job_config,
+                                     job_id_prefix=f'clean_{module_name}_')
+
+            # wait for job to complete
+            query_job.result()
```

Comment:
A job may complete but have job.errors. Please raise an exception here and ensure we log it and the rule_info, query_dic, etc. below.

Reply:
Ooh also include the `job_id`!